src/psyclone/psyir/transformations/omp_loop_trans.py

```diff
@@ -131,6 +138,59 @@ def __init__(self, omp_directive="do", omp_schedule="auto"):
     def __str__(self):
         return "Adds an OpenMP directive to parallelise the target loop"
 
+    def _add_asynchronicity(self, node: Loop, instance: Directive):
+        '''
+        TODO
+        '''
+        # Only OMPDoDirective supports nowait of these directives. Needs to be
+        # exact type check
+        if type(instance) is not OMPDoDirective:
+            return
+        # Otherwise find the next dependency.
+        next_depend = self._find_next_dependency(node, instance)
+        # If find_next_dependency returns False, then this loop is its own
+        # next dependency so we can't add an asynchronous clause.
+        if not next_depend:
+            return
+
+        barrier_type = MAP_STR_TO_BARRIER_DIRECTIVE[self.omp_directive]
+        # If find next_dependency returns True there is no follow up
+        # dependency, so we just need a barrier at the end of the containing
+        # Routine.
+        if next_depend is True:
+            # Add nowait to the instance.
+            instance.nowait = True
+            # Add a barrier to the end of the containing Routine if there
+            # isn't one already.
+            containing_routine = node.ancestor((Routine, OMPParallelDirective))
+            containing_schedule = containing_routine.walk(Schedule)[0]
+            # Check barrier that corresponds to self.omp_directive and add the
+            # correct barrier type
+            if not isinstance(containing_schedule.children[-1], barrier_type):
+                containing_schedule.addchild(barrier_type())
+            return
+
+        # Otherwise we have the next dependency and we need to find where the
+        # correct place to place the preceding barrier is. Need to find a
+        # guaranteed control flow path to place it.
+
+        # Find the deepest schedule in the tree containing both.
+        sched = next_depend.ancestor(Schedule)
+        routine = node.ancestor(Routine)
```

Comment:
Can you simplify by using the `shared_with` argument to `ancestor`?

Reply:
It seems so.